rust/lance/src/catalog/catalog_trait.rs

```diff
@@ -0,0 +1,276 @@
+    async fn list_tables(
+        &self,
+        start_after: Option<String>,
+        limit: Option<u32>,
+    ) -> Result<Vec<String>>;
+}
+
+/// A simplified catalog that puts all tables in a single base directory
+///
+/// The object store's CAS primitives are used for commit handling.
+///
+/// This object store is simplistic but has zero dependencies (beyond an object store
+/// of some kind)
+#[derive(Debug)]
+pub struct ListingCatalog {
+    base_path: Path,
+    object_store: ObjectStore,
+    commit_handler: Arc<dyn CommitHandler>,
+}
+
+const LANCE_EXTENSION: &str = "lance";
+
+fn format_table_url_or_path(is_url: bool, scheme: &str, base_path: &str, name: &str) -> String {
+    if is_url {
+        format!("{}:///{}/{}.{}", scheme, base_path, name, LANCE_EXTENSION)
+    } else {
+        format!("{}/{}.{}", base_path, name, LANCE_EXTENSION)
+    }
+}
+
+impl ListingCatalog {
+    pub fn new(base_path: Path, object_store: ObjectStore) -> Self {
+        Self {
+            base_path,
+            object_store: object_store.clone(),
+            commit_handler: Arc::new(ListingCommitHandler { object_store }),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Catalog for ListingCatalog {
+    async fn create_table(&self, name: &str, schema: Arc<Schema>) -> Result<TableReference> {
+        let table_url = format_table_url_or_path(
+            true,
+            &self.object_store.scheme,
+            self.base_path.as_ref(),
+            name,
+        );
+
+        let ds = Dataset::write(
+            RecordBatchIterator::new(vec![].into_iter().map(Ok), schema.clone()),
+            &table_url,
+            Some(WriteParams {
+                commit_handler: Option::from(self.commit_handler.clone()),
+                store_params: Some(ObjectStoreParams {
+                    object_store: Some((
+                        self.object_store.clone().inner,
+                        Url::parse(&table_url).unwrap(),
+                    )),
+                    ..Default::default()
+                }),
+                ..Default::default()
+            }),
+        )
+        .await?;
+
+        Ok(TableReference {
+            uri: ds.uri().to_string(),
+            object_store: self.object_store.clone(),
+            commit_handler: self.commit_handler.clone(),
+            store_params: Some(ObjectStoreParams {
+                object_store: Some((
+                    self.object_store.clone().inner,
+                    Url::parse(&table_url).unwrap(),
+                )),
+                ..Default::default()
+            }),
+        })
+    }
+
+    async fn drop_table(&self, name: &str) -> Result<()> {
+        let table_path = format_table_url_or_path(
+            false,
+            &self.object_store.scheme,
+            self.base_path.as_ref(),
+            name,
+        );
+
+        self.object_store.remove_dir_all(table_path).await?;
```

Comment:
Is this `remove_dir_all`  executing atomically ?   After skimmed the implementation, I think it's not, since I see it list firstly, and then delete the entries one by one.  So I think it may have consistent issues.

Reply:
Yes, it's not an atomical operation. It only refers to the `drop` api's implementation for the demo purpose.